src/cryptsetup/cryptsetup-tokens/cryptsetup-token-systemd-fido2.c

```diff
@@ -0,0 +1,231 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <libcryptsetup.h>
+
+#include "cryptsetup-token.h"
+#include "cryptsetup-token-util.h"
+#include "hexdecoct.h"
+#include "json.h"
+#include "luks2-fido2.h"
+#include "memory-util.h"
+#include "strv.h"
+
+#define TOKEN_NAME "systemd-fido2"
+#define TOKEN_VERSION_MAJOR "1"
+#define TOKEN_VERSION_MINOR "0"
+
+/* for libcryptsetup debug purpose */
+_public_ const char *cryptsetup_token_version(void) {
+        return TOKEN_VERSION_MAJOR "." TOKEN_VERSION_MINOR;
+}
+
+_public_ int cryptsetup_token_open_pin(
+                struct crypt_device *cd, /* is always LUKS2 context */
+                int token /* is always >= 0 */,
+                const char *pin,
+                size_t pin_size,
+                char **password, /* freed by cryptsetup_token_buffer_free */
+                size_t *password_len,
+                void *usrptr /* plugin defined parameter passed to crypt_activate_by_token*() API */) {
+
+        int r;
+        const char *json;
+        _cleanup_strv_free_erase_ char **pins = NULL;
+
+        assert(token >= 0);
+        assert(password);
+        assert(password_len);
+
+        /* This must not fail at this moment (internal error) */
+        r = crypt_token_json_get(cd, token, &json);
+        assert(token == r);
+        assert(json);
+
+        /* systemd fido2 code uses pin lists internally, libcryptsetup does not */
+        if (pin) {
+                if (strnlen(pin, pin_size) != pin_size ||
+                    pin[pin_size] != '\0') {
```

Comment:
Well...that was pretty stupid mistake on my side. I've changed it so that the check (which was wrong) was not needed for pkcs11 plugin at all.

For fido2 it's a bit more complicated:
The libcryptsetup API expects pin parameter to by arbitrary buffer delimited by pin_size.
Fido2 systemd internals expects pin to be char string always. For it to work correctly I need to double check that pin buffer is always terminated by NULL byte. The correct way is to pass whole string buffer including the terminating NULL and reflect that in pin_size parameter. OTOH, it's quite inconvenient when usually applications pass couple: pin, strlen(pin).
When they do I'll just copy the pin in new buffer terminated by NULL by using strNdup() now.